### PR TITLE
[data] Remove unification of schemas

### PR DESCRIPTION
**Pull request by iamjustinhsu**
## Why are these changes needed?
- unification of schemas is slow
- revert back to pre https://github.com/ray-project/ray/pull/53454/files commit. if no unification before, no unification after. if unification before, we can leave it there or add it back. If I removed it I added a comment with `# NOTE`

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/block.py [hunk 1]
```diff
@@ -114,6 +114,21 @@ def _is_empty_schema(schema: Optional[Schema]) -> bool:
...
+    """Return the first non-empty schema from an iterator of schemas.
+
+    Args:
+        schemas: Iterator of schemas to check.
+
+    Returns:
+        The first non-empty schema, or None if all schemas are empty.
+    """
+    for schema in schemas:
+        if not _is_empty_schema(schema):
```
Comment (alexeykudinkin) [inline python/ray/data/block.py:127]: Why would there be empty schemas?

File: python/ray/data/context.py [hunk 2]
```diff
@@ -537,6 +539,8 @@ class DataContext:
     downstream_capacity_backpressure_ratio: float = None
     downstream_capacity_backpressure_max_queued_bundles: int = None
 
+    allow_divergent_schemas: bool = DEFAULT_ALLOW_DIVERGENT_SCHEMAS
```
Comment (alexeykudinkin) [inline python/ray/data/context.py]: Let's rename to `enforce_schemas` (along with util params)

File: python/ray/data/block.py (hunk 1 above)
Comment (iamjustinhsu) [reply python/ray/data/block.py:127]: I'm not confident enough to remove this statement because we know empty schemas have appeared before

Comment (iamjustinhsu) [reply python/ray/data/block.py:127]: ok my most recent commit found a case [9734268](https://github.com/ray-project/ray/pull/55926/commits/9734268902f960bc7496c32aa4dc3a7857d0a0ac) where we are taking the first non-empty bundle

File: python/ray/data/_internal/planner/aggregate.py [hunk 3]
```diff
@@ -52,6 +52,7 @@ def fn(
             metadata.extend(ref_bundle.metadata)
         if len(blocks) == 0:
             return (blocks, {})
+        # NOTE: unify was here before.
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/planner/aggregate.py]: Don't forget to drop these

File: python/ray/data/_internal/execution/streaming_executor_state.py [hunk 4]
```diff
@@ -793,7 +796,7 @@ def dedupe_schemas_with_validation(
             f"than the previous one. Previous schema: {old_schema}, "
             f"new schema: {bundle.schema}. This may lead to unexpected behavior."
         )
-    if allow_divergent:
+    if not enforce_schemas:
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/execution/streaming_executor_state.py]: This should be the opposite -- we should reconcile only if we enforce schemas

File: python/ray/data/context.py [hunk 5]
```diff
@@ -140,6 +140,8 @@ class ShuffleStrategy(str, enum.Enum):
     "RAY_DATA_ENABLE_PROGRESS_BAR_NAME_TRUNCATION", True
 )
 
+DEFAULT_ENFORCE_SCHEMAS = env_bool("RAY_DATA_ALLOW_ENFORCE_SCHEMAS", True)
```
Comment (alexeykudinkin) [inline python/ray/data/context.py]: Should be false by default

**Review by alexeykudinkin — COMMENTED** [review]
@iamjustinhsu make sure we add a regression test for Attentive

File: python/ray/data/tests/test_unify_schemas_performance.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        (_create_simple_schema, 0.00001),
+        (_create_tensor_fixed_schema, 0.00005),
+        (_create_tensor_variable_schema, 0.00005),
+        (_create_object_schema, 0.00005),
+        (_create_nested_struct_schema, 0.0001),
+        (_create_deep_nested_schema, 0.0002),
+        (_create_mixed_complex_schema, 0.0002),
+    ],
+)
+def test_unify_schemas_equivalent_performance(
```
Comment (iamjustinhsu) [inline python/ray/data/tests/test_unify_schemas_performance.py:96]: @goutamvenkat-anyscale should handle cases where the schemas are NOT identical

File: python/ray/data/tests/test_deduping_schema.py [hunk 7]
```diff
@@ -7,36 +7,46 @@
...
+        (pa.schema([pa.field("id", pa.int64())]), False),
+        (pa.schema([]), True),  # Empty Schema
+        (PandasBlockSchema(names=["col2"], types=[int]), False),
+        (PandasBlockSchema(names=[], types=[]), True),
+        (None, True),  # Null Schema
     ],
 )
 def test_dedupe_schema_handle_empty(
     old_schema: Optional["Schema"],
+    old_schema_is_empty: bool,
```
Comment (alexeykudinkin) [inline python/ray/data/tests/test_deduping_schema.py]: Why do we need this flag? You can just test if the schema is empty, right?

Comment (iamjustinhsu) [reply python/ray/data/tests/test_deduping_schema.py]: @alexeykudinkin  yea it's not completely necessary, just thought it would be easier to read than before, i can revert back

**Review by alexeykudinkin — APPROVED** [review]

**Comment by aslonnie** [comment]
cc @aslonnie
